### PR TITLE
note some safety concerns of raw-ptr-to-ref casts

### PR DESCRIPTION
**Pull request by RalfJung**

### CODE REVIEW

File: src/libcore/ptr/mod.rs [hunk 1]
```diff
@@ -1759,6 +1777,13 @@ impl<T: ?Sized> *mut T {
     /// of the returned pointer, nor can it ensure that the lifetime `'a`
     /// returned is indeed a valid lifetime for the contained data.
     ///
+    /// When calling this method, you have to ensure that if the pointer is
+    /// non-NULL, then it is properly aligned, dereferencable (for the whole
+    /// size of `T`) and points to an initialized instance of `T`. This applies
+    /// even if the result of this method is unused!
+    /// (The part about being initialized is not yet fully decided, but until
+    /// it is the only safe approach is to ensure that they are indeed initialized.)
+    ///
```
Comment (rkruppe) [inline src/libcore/ptr/mod.rs]: I think there should be an equivalent of `as_ref`'s paragraph about the lifetime here (responsibility to ensure there are no other references).

File: src/libcore/ptr/mod.rs [hunk 2]
```diff
@@ -984,8 +984,17 @@ impl<T: ?Sized> *const T {
     /// operation because the returned value could be pointing to invalid
     /// memory.
     ///
+    /// When calling this method, you have to ensure that if the pointer is
+    /// non-NULL, then it is properly aligned, dereferencable (for the whole
+    /// size of `T`) and points to an initialized instance of `T`. This applies
+    /// even if the result of this method is unused!
+    /// (The part about being initialized is not yet fully decided, but until
+    /// it is the only safe approach is to ensure that they are indeed initialized.)
```
Comment (rkruppe) [inline src/libcore/ptr/mod.rs]:
```suggestion
    /// it is, the only safe approach is to ensure that they are indeed initialized.)
```

File: src/libcore/ptr/mod.rs [hunk 3]
```diff
@@ -1759,6 +1777,13 @@ impl<T: ?Sized> *mut T {
     /// of the returned pointer, nor can it ensure that the lifetime `'a`
     /// returned is indeed a valid lifetime for the contained data.
     ///
+    /// When calling this method, you have to ensure that if the pointer is
+    /// non-NULL, then it is properly aligned, dereferencable (for the whole
+    /// size of `T`) and points to an initialized instance of `T`. This applies
+    /// even if the result of this method is unused!
+    /// (The part about being initialized is not yet fully decided, but until
+    /// it is the only safe approach is to ensure that they are indeed initialized.)
```
Comment (rkruppe) [inline src/libcore/ptr/mod.rs]:
```suggestion
    /// it is, the only safe approach is to ensure that they are indeed initialized.)
```

**Review by hanna-kruppe — COMMENTED** [review]
LGTM, but some nits.

File: src/libcore/ptr/mod.rs (hunk 1 above)
Comment (RalfJung) [reply src/libcore/ptr/mod.rs]: I added a paragraph about this, does that look good to you?

Comment (rkruppe) [reply src/libcore/ptr/mod.rs]: 👍 r=me after squashing

**Pull request merged**